File: solutions/230.kth-smallest-element-in-a-bst.java
/*
 * @lc app=leetcode id=230 lang=java
 *
 * [230] Kth Smallest Element in a BST
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        List<Integer> list = new ArrayList<>();
        visitNodes(root, list);
        return list.get(k - 1);
    }

    private void visitNodes(TreeNode node, List<Integer> list) {
        if (node == null) return;

        visitNodes(node.left, list);
        list.add(node.val);
        visitNodes(node.right, list);
    }

    public int kthSmallest_dfs(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode node = root;
        while (!stack.isEmpty() || node != null) {
            while (node != null) {
                stack.add(node);
                node = node.left;
            }
            node = stack.pop();
            if (--k == 0) return node.val;
            node = node.right;
        }
        return 0;
    }
}
// @lc code=end

File: solutions/230.kth-smallest-element-in-a-bst.py
#
# @lc app=leetcode id=230 lang=python3
#
# [230] Kth Smallest Element in a BST
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def inOrderTraversal(node):
            if not node:
                return
            yield from inOrderTraversal(node.left)
            yield node.val
            yield from inOrderTraversal(node.right)

        gen = inOrderTraversal(root)
        for _ in range(k):
            result = next(gen)
        return result

    def kthSmallest_rec(self, root: Optional[TreeNode], k: int) -> int:
        result = []

        def process(node):
            nonlocal result
            if not node:
                return
            process(node.left)
            result.append(node.val)
            process(node.right)

        process(root)
        return result[k - 1]

    def kthSmallest_dfs(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        current = root
        while stack or current:
            while current:
                stack.append(current)
                current = current.left
            current = stack.pop()
            k -= 1
            if k == 0:
                return current.val
            current = current.right

        
# @lc code=end

